src/DeviceInput/InputDevices/deviceSourceManager.ts

```diff
@@ -0,0 +1,308 @@
+
+    // Private Members
+    private _devices: Array<Array<DeviceSource<DeviceType>>>;
+    private _firstDevice: Array<number>;
+    private _deviceInputSystem: DeviceInputSystem;
+
+    /**
+     * Default Constructor
+     * @param engine engine to pull input element from
+     * @param enableObserveEvents boolean to enable use of observe events
+     */
+    constructor(engine: Engine, enableObserveEvents: boolean = false) {
+        const numberOfDeviceTypes = Object.keys(DeviceType).length / 2;
+        this._devices = new Array<Array<DeviceSource<DeviceType>>>(numberOfDeviceTypes);
+        this._firstDevice = new Array<number>(numberOfDeviceTypes);
+        this._deviceInputSystem = new DeviceInputSystem(engine);
+
+        this._deviceInputSystem.onDeviceConnected = (deviceName) => {
+            const deviceType = this._getDeviceTypeFromName(deviceName);
+            const deviceSlot = this._getDeviceSlot(deviceName);
+
+            this.onBeforeDeviceConnectedObservable.notifyObservers({ deviceType, deviceSlot });
+            this._addDevice(deviceName);
+            this.onAfterDeviceConnectedObservable.notifyObservers({ deviceType, deviceSlot });
+        };
+        this._deviceInputSystem.onDeviceDisconnected = (deviceName) => {
+            const deviceType = this._getDeviceTypeFromName(deviceName);
+            const deviceSlot = this._getDeviceSlot(deviceName);
+
+            this.onBeforeDeviceDisconnectedObservable.notifyObservers({ deviceType, deviceSlot });
+            this._removeDevice(deviceName);
+            this.onAfterDeviceDisconnectedObservable.notifyObservers({ deviceType, deviceSlot });
+        };
+
+        if (!this._deviceInputSystem.observeInput && enableObserveEvents) {
+            this._deviceInputSystem.observeInput = (deviceName, inputIndex, previousState, currentState) => {
+                const deviceType = this._getDeviceTypeFromName(deviceName);
+                const deviceSlot = this._getDeviceSlot(deviceName);
+
+                if (deviceType == DeviceType.Keyboard || deviceType == DeviceType.Mouse || deviceType == DeviceType.Touch) {
+                    this.getDeviceSource(deviceType, deviceSlot)?.onInputChangedObservable.notifyObservers({ inputIndex, previousState, currentState });
+                }
+            };
+        }
+    }
+
+    // Public Functions
+    /**
+     * Checks for current device input value, given DeviceType, slot, and inputIndex
+     * @param type Enum specifiying device type
+     * @param inputIndex Index of device input
+     * @param deviceSlot "Slot" or index that device is referenced in
+     * @returns Current value of input
+     */
+    public getInput<T extends DeviceType>(type: T, inputIndex: DeviceInputs<T>, deviceSlot: number = this._firstDevice[type]): Nullable<number> {
+        if (!this._devices[type] || this._firstDevice[type] === undefined || this._devices[type][deviceSlot] === undefined) {

```

Comment:
There are times when this will be undefined, such as when it's checking for a gamepad that isn't connected.

Reply:
Does that means deviceSlot can be `undefined`?

Reply:
Yes.  If all devices for a given DeviceType are removed, deviceSlot could be undefined.

Reply:
Okay, so deviceSlot currently has a type of number but it can also be undefined. Maybe you should default deviceSlot to undefined and set the deviceSlot manually after checking the _firstDevice.